pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -377,111 +369,112 @@ private boolean isUpsertPreloadEnabled() {
+    IndexLoadingConfig indexLoadingConfig = getIndexLoadingConfig(zkMetadata);
+    handleUpsertPreload(zkMetadata, indexLoadingConfig);
     SegmentDataManager segmentDataManager = _segmentDataManagerMap.get(segmentName);
-    if (segmentDataManager != null) {
-      if (upsertPreloadEnabled) {
-        _logger.debug(
-            "Skipping adding existing segment: {} for table: {} with data manager class: {}, as it's preloaded",
-            segmentName, _tableNameWithType, segmentDataManager.getClass().getSimpleName());
+    if (segmentDataManager == null) {
+      addNewOnlineSegment(zkMetadata, indexLoadingConfig);
+    } else {
+      if (segmentDataManager instanceof RealtimeSegmentDataManager) {
+        _logger.info("Changing segment: {} from CONSUMING to ONLINE", segmentName);
+        ((RealtimeSegmentDataManager) segmentDataManager).goOnlineFromConsuming(zkMetadata);
+        onConsumingToOnline(segmentName);
       } else {
-        _logger.warn("Skipping adding existing segment: {} for table: {} with data manager class: {}", segmentName,
-            _tableNameWithType, segmentDataManager.getClass().getSimpleName());
+        replaceSegmentIfCrcMismatch(segmentDataManager, zkMetadata, indexLoadingConfig);
       }
+    }
+  }
+
+  @Override
+  public void addConsumingSegment(String segmentName) {
+    Preconditions.checkState(!_shutDown,
+        "Table data manager is already shut down, cannot add CONSUMING segment: %s to table: %s", segmentName,
+        _tableNameWithType);
+    _logger.info("Adding CONSUMING segment: {}", segmentName);
+    Lock segmentLock = getSegmentLock(segmentName);
+    segmentLock.lock();
+    try {
+      doAddConsumingSegment(segmentName);
+    } catch (Exception e) {
+      addSegmentError(segmentName,
+          new SegmentErrorInfo(System.currentTimeMillis(), "Caught exception while adding CONSUMING segment", e));
+      throw e;
+    } finally {
+      segmentLock.unlock();
+    }
+  }
+
+  private void doAddConsumingSegment(String segmentName) {
+    SegmentZKMetadata zkMetadata = getZKMetadata(segmentName);
+    // NOTE: We do not throw exception here because the segment might have just been committed before the state
+    //       transition is processed. We can skip adding this segment, and rely on the following CONSUMING -> ONLINE
+    //       state transition to add it.
+    if (zkMetadata.getStatus() != Status.IN_PROGRESS) {
+      _logger.warn("Segment: {} is already committed, skipping adding it as CONSUMING segment", segmentName);
+      return;
+    }
+    IndexLoadingConfig indexLoadingConfig = getIndexLoadingConfig(zkMetadata);
+    handleUpsertPreload(zkMetadata, indexLoadingConfig);
+    SegmentDataManager segmentDataManager = _segmentDataManagerMap.get(segmentName);
+    if (segmentDataManager != null) {
+      _logger.warn("Segment: {} already exists, skipping adding it as CONSUMING segment", segmentName);
```

Comment:
nit: check if segmentDataManager is mutable/immutable segment and log it for debug